fuzzers/main.py

```diff
@@ -0,0 +1,842 @@
+
+        This is a synchronous operation for convenience reasons, as cargo itself uses all the
+        available cores most of the time it shouldn't be a big deal. The only drawback is that
+        requests to pause/exit the fuzzer would block until the current build is completed.
+        """
+        print(f'Building fuzzer for branch {self.branch} and target '
+              f'{self.target}, log is at {self.log_relpath}', file=sys.stderr)
+
+        # Log metadata information
+        current_commit = str(
+            subprocess.run(
+                ['git', 'rev-parse', 'HEAD'],
+                cwd=self.repo_dir,
+                capture_output=True,
+                check=True,
+            ).stdout)
+        self.log_file.write(f'Corpus version: {self.corpus_vers}\n')
+        self.log_file.write(
+            f'On commit {current_commit} (tip of branch {self.branch["name"]})\n'
+        )
+        self.log_file.write(f'Target is {self.target}\n')
+        self.log_file.write(f'Current time: {datetime.datetime.now()}\n')
+        self.log_file.flush()
+
+        # Build the fuzzer runner
+        # TODO: actually build and have build time metrics for the fuzzer runner, but this must not
+        # block the pause/resume behavior
+        #build_start = time.monotonic()
+        #subprocess.check_call(
+        #    ['cargo', 'fuzz', 'build', self.target['runner']],
+        #    cwd=self.repo_dir / self.target['crate'],
+        #    stdout=self.log_file,
+        #    stderr=subprocess.STDOUT,
+        #)
+        #self.fuzz_build_time_metric.inc(time.monotonic() - build_start)
+
+    def start(self, corpus: Corpus) -> None:
+        """Start the fuzzer runner on corpus `Corpus`"""
+        print(f'Starting fuzzer for branch {self.branch} and '
+              f'target {self.target}, log is at {self.log_relpath}',
+              file=sys.stderr)
+
+        # Prepare the fuzz time metric
+        self.last_time = time.monotonic()
+
+        # Spin up the fuzzer process itself
+        # libfuzzer will kill the process if it takes more than -timeout number of seconds.
+        # nayduck can sigstop the fuzzing process for ~2 hours at most, so 8000s should be ok.
+        self.proc = subprocess.Popen(  # pylint: disable=consider-using-with
+            [
+                'cargo',
+                'fuzz',
+                'run',
+                self.target['runner'],
+                '--',
+                str(corpus.corpus_for(self.target)),
+                str(corpus.artifacts_for(self.target)),
+                f'-artifact_prefix={corpus.artifacts_for(self.target)}/',
+                '-timeout=8000',
+            ] + self.target['flags'],
+            cwd=self.repo_dir / self.target['crate'],
+            start_new_session=True,
+            stdout=self.log_file,
+            stderr=subprocess.STDOUT,
+        )
+
+    def poll(self) -> bool:
+        """Checks if the current process is still running. Returns True if it stopped"""
```

Comment:
That’s a bit confusing name / return value. How about `is_running`?

Reply:
Right, but Popen.poll returns exit code or None so that’s why calling this method poll may be confusing.

Maybe just `check_is_running`?